A29-xds-tls-security.md

```diff
@@ -0,0 +1,422 @@
+[`FilterChainMatch`][filter-chain-match]. The [`transport_socket`][transport-socket]
+of the matched (selected) `filter_chain` is used to extract the
+[`DownstreamTlsContext`][DTC] as described [here][transport-socket-comment].
+The `DownstreamTlsContext` thus obtained is used for the incoming connection.
+
+[`common_tls_context`][CTC1] in the `DownstreamTlsContext` contains the required
+security configuration. The [`require_client_certificate`][RCC] field determines
+whether the client certificate is required (mTLS mode if `true`, TLS if `false`).
+All the other fields are ignored. See below for [`CommonTlsContext`][CTC-type]
+processing.
+
+[A36]: A36-xds-for-servers.md
+[filter-chains]: https://github.com/envoyproxy/envoy/blob/45ec050f91407147ed53a999434b09ef77590177/api/envoy/config/listener/v3/listener.proto#120
+[filter-chain-match]: A36-xds-for-servers.md#filterchainmatch
+[transport-socket]: https://github.com/envoyproxy/envoy/blob/45ec050f91407147ed53a999434b09ef77590177/api/envoy/config/listener/v3/listener_components.proto#L246
+[DTC]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L57
+[transport-socket-comment]: https://github.com/envoyproxy/envoy/blob/45ec050f91407147ed53a999434b09ef77590177/api/envoy/config/listener/v3/listener_components.proto#L241
+[CTC1]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#84
+[RCC]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#88
+[CTC-type]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L129
+
+#### CommonTlsContext Processing
+
+`CommonTlsContext` is present in both the `UpstreamTlsContext` and `DownstreamTlsContext` protos and contains the
+certificate and key configuration information needed on the client and server side respectively. The configuration
+tells gRPC how to obtain certificates and the keys for the TLS handshake. Although there are various ways to obtain
+certificates as per this proto (which are supported by Envoy), gRPC supports only one of them and that is defined by the
+[`CertificateProviderInstance`][CPI] proto which is based on the notion of CertificateProvider plugin framework
+(described later). The field `instance_name` defines a CertificateProvider "instance" that gRPC looks up in the
+bootstrap file (described later) to obtain the CertificateProvider configuration. This configuration along with
+the CertificateProvider plugin framework enables gRPC to acquire the certificates necessary for the TLS handshake.
+
+The field [`tls_certificate_certificate_provider_instance`][TCCPI] is used for the local certificate and the private
+key. And either [`validation_context_certificate_provider_instance`][VCCPI] or
```

Comment:
As discussed, I agree with you. I'll change the text to only support `validation_context_certificate_provider_instance` inside of `combined_validation_context`.

Reply:
I was talking with @ejona86 about this just last week.  I think we misunderstood how the xDS validation config worked when we added those fields.  The [`validation_context_certificate_provider`](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L255) and [`validation_context_certificate_provider_instance`](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L259) fields don't actually make any sense, because cert providers don't return a complete [`CertificateValidationContext` proto](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L236); they just provide the CA certs, which are basically just the [`trusted_ca` field](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L282) within that proto.

What we probably should have done was simply add the new cert provider fields in `CommonTlsContext` but not inside of the [`validation_context_type` oneof](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L236).  The idea would be that the `CertificateValidationContext` is obtained from one of [`validation_context`](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L238), [`validation_context_sds_secret_config`](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L242), or [`combined_validation_context`](https://github.com/envoyproxy/envoy/blob/ea063c0dfebe57877de7968d51b650c1991a1c8d/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L251), fully populated *except* for the `trusted_ca` field.  Then, as a second step, the cert provider is used to populate the `trusted_ca` field.

I've put together envoyproxy/envoy#17201 for this.  We'll need to update all 3 languages and TD.

CC @jaychenatr

Reply:
I think the open issue of potential changes to the CertificateProvider fields in xDS is the biggest open issue in this document.  We really need to get this nailed down before we finalize this gRFC.